File: apple/Cabalmail/ViewModels/FeedItemDetailViewModel.swift
import Foundation
import Observation
import CabalmailKit

/// Backs `FeedItemDetailView`: which surface is showing (the in-feed body or
/// the publisher's article), the styling toggle, remote content, and the
/// item's state.
///
/// The initial surface, styling, and remote-content policy come from the
/// subscription's per-feed preferences (`default_open_mode`,
/// `default_styling`, `default_remote_content`), which the server stores and
/// the client applies (open Q5 of the requirements). The toolbar toggles are
/// **sticky per feed**: flipping one writes it back as that feed's default,
/// so the next item in the feed opens the same way — the settings sheet's
/// pickers and the reader's toggles are two views of the same three fields.
@Observable
@MainActor
final class FeedItemDetailViewModel {
    var item: RssItem
    /// The feed's settings row, kept current with every sticky write so a
    /// second toggle compares against what the first one stored.
    private(set) var subscription: RssSubscription?
    /// True while the publisher's page is on screen instead of the body.
    var showingArticle: Bool
    /// Reader styling (the mail reader's stylesheet for the body; Readability
    /// for the article) versus the author's own.
    var readerMode: Bool
    var remoteContentAllowed: Bool

    private let engine: RssSyncEngine?
    private let defaults: (any FeedDefaultsPersisting)?
    private let bus: FeedStateBus

    init(
        item: RssItem,
        subscription: RssSubscription?,
        engine: RssSyncEngine?,
        preferences: Preferences,
        defaults: (any FeedDefaultsPersisting)? = nil,
        bus: FeedStateBus = .shared
    ) {
        self.item = item
        self.subscription = subscription
        self.engine = engine
        // The sync engine is the production persister; tests hand in a fake.
        self.defaults = defaults ?? engine
        self.bus = bus
        let policy = FeedDetailPolicy.initial(
            for: subscription,
            hasArticleURL: URL(string: item.url) != nil,
            globalRemoteContent: preferences.loadRemoteContent
        )
        self.showingArticle = policy.showsArticle
        self.readerMode = policy.readerMode
        self.remoteContentAllowed = policy.remoteContentAllowed
        bus.subscribe(self) { [weak self] change in self?.apply(change) }
    }

    /// The list's swipe or context menu changed this item: keep the toolbar
    /// truthful.
    func apply(_ change: RssItem?) {
        guard let change, change.id == item.id else { return }
        item.isRead = change.isRead
        item.isFavorite = change.isFavorite
    }

    var articleURL: URL? {
        guard let url = URL(string: item.url), url.scheme?.lowercased() == "https" || url.scheme?.lowercased() == "http"
        else { return nil }
        return url
    }

    /// The per-subscription web-view storage identifier (D11): cookies and
    /// local storage set on the publisher's site stay with this feed.
    var dataStoreID: UUID? {
        subscription.flatMap { UUID(uuidString: $0.dataStoreUuid) }
    }

    func toggleArticle() {
        showingArticle.toggle()
        persistDefault(.article(showing: showingArticle))
    }

    func toggleReaderMode() {
        readerMode.toggle()
        persistDefault(.styling(readerMode: readerMode))
    }

    func toggleRemoteContent() {
        remoteContentAllowed.toggle()
        persistDefault(.remoteContent(allowed: remoteContentAllowed))
    }

    /// Writes the toggled choice back as the feed's default if it differs
    /// from what is stored — only that one field, so flipping the article
    /// view never pins remote content as a side effect. Optimistic: the
    /// in-memory row moves at once (the persister does the same for the
    /// local store, so the next item opened in this feed already honours
    /// it); the server round trip is best-effort and a failure is reconciled
    /// by the next catalog refresh. Nothing to write without a known
    /// subscription.
    private func persistDefault(_ toggle: FeedReaderToggle) {
        guard let subscription, let defaults,
              let update = FeedDetailPolicy.stickyUpdate(
                  for: subscription, toggle: toggle, hasArticleURL: articleURL != nil
              )
        else { return }
        self.subscription = subscription.applying(update)
        Task { [bus] in
            _ = try? await defaults.updateSubscription(subscription, update)
            // The sidebar and settings sheet read the row too.
            bus.postCatalogChanged()
        }
    }

    func setRead(_ isRead: Bool) async {
        item.isRead = isRead
        try? await engine?.setRead(item, isRead)
        bus.post(item)
    }

    func setFavorite(_ isFavorite: Bool) async {
        item.isFavorite = isFavorite
        try? await engine?.setFavorite(item, isFavorite)
        bus.post(item)
    }
}

/// Where the reader's sticky per-feed defaults are written. `RssSyncEngine`
/// is the production implementation (local store first, then the server);
/// tests substitute a recorder.
protocol FeedDefaultsPersisting: Sendable {
    func updateSubscription(_ subscription: RssSubscription, _ update: RssSubscriptionUpdate) async throws
        -> RssSubscription
}

extension RssSyncEngine: FeedDefaultsPersisting {}

/// What the reader shows first for an item, from the subscription's stored
/// preferences, and what a toggle writes back. Pure so it can be unit-tested.
enum FeedDetailPolicy {
    struct Initial: Equatable {
        var showsArticle: Bool
        var readerMode: Bool
        var remoteContentAllowed: Bool = false
    }

    static func initial(
        for subscription: RssSubscription?,
        hasArticleURL: Bool,
        globalRemoteContent: LoadRemoteContentPolicy = .off
    ) -> Initial {
        let openMode = subscription?.defaultOpenMode ?? .summary
        let styling = subscription?.defaultStyling ?? .reader
        let remote: Bool
        switch subscription?.defaultRemoteContent ?? .inherit {
        case .show: remote = true
        case .hide: remote = false
        // "Ask" is the mail reader's per-message prompt; the feed reader has
        // no prompt, so only an explicit Always shows remote content by
        // default. Same rule as before the per-feed override existed.
        case .inherit: remote = globalRemoteContent == .always
        }
        // "Article" can only be honoured when the item links somewhere.
        return Initial(
            showsArticle: openMode == .article && hasArticleURL,
            readerMode: styling == .reader,
            remoteContentAllowed: remote
        )
    }

    /// The one field a reader toggle would change on the feed's stored
    /// defaults, or nil when it already matches. The open mode is only
    /// written when the item has an article to show — without one
    /// `showsArticle` is false regardless of preference, and writing that
    /// back would silently revert an "Article" default. Remote content is
    /// written as an explicit `show` / `hide`: a toggle is a decision about
    /// this feed, which is exactly what `inherit` isn't.
    static func stickyUpdate(
        for subscription: RssSubscription,
        toggle: FeedReaderToggle,
        hasArticleURL: Bool
    ) -> RssSubscriptionUpdate? {
        var update = RssSubscriptionUpdate()
        switch toggle {
        case .article(let showing):
            guard hasArticleURL else { return nil }
            let openMode: RssOpenMode = showing ? .article : .summary
            if openMode != subscription.defaultOpenMode { update.defaultOpenMode = openMode }
        case .styling(let readerMode):
            let styling: RssStyling = readerMode ? .reader : .native
            if styling != subscription.defaultStyling { update.defaultStyling = styling }
        case .remoteContent(let allowed):
            let remote: RssRemoteContentMode = allowed ? .show : .hide
            if remote != subscription.defaultRemoteContent { update.defaultRemoteContent = remote }
        }
        return update.isEmpty ? nil : update
    }
}

/// A reader toolbar toggle and the state it left behind.
enum FeedReaderToggle: Equatable {
    case article(showing: Bool)
    case styling(readerMode: Bool)
    case remoteContent(allowed: Bool)
}

File: android/app/src/main/kotlin/com/cabalmail/android/ui/feeds/FeedItemDetailViewModel.kt
package com.cabalmail.android.ui.feeds

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import com.cabalmail.android.AppContainer
import com.cabalmail.kit.models.RssItem
import com.cabalmail.kit.models.RssSubscription
import com.cabalmail.kit.models.RssSubscriptionUpdate
import com.cabalmail.kit.rss.RssStore
import com.cabalmail.kit.rss.RssSyncEngine
import com.cabalmail.kit.settings.AppPreferences
import com.cabalmail.kit.settings.MarkAsRead
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class FeedItemDetailUiState(
    val item: RssItem? = null,
    val subscription: RssSubscription? = null,
    val showingArticle: Boolean = false,
    val readerMode: Boolean = true,
    val remoteContentAllowed: Boolean = false,
    val loaded: Boolean = false,
    val error: String? = null,
) {
    /** The article link, when it is a web address. */
    val articleUrl: String?
        get() =
            item?.url?.takeIf {
                val lower = it.lowercase()
                lower.startsWith("http://") || lower.startsWith("https://")
            }
}

/**
 * One feed item's reader, the Apple `FeedItemDetailViewModel`. The item
 * and its subscription are read from the store at creation, and the
 * feed's defaults (open mode, styling, remote content) decide the first
 * render; the three toggles write back to the feed's row so the next item
 * opens the same way, optimistic in-memory first, the server round trip
 * best-effort. Mark-read-on-open follows the feed reader's own
 * preference.
 */
class FeedItemDetailViewModel(
    private val feedId: String,
    private val sortKey: String,
    private val store: RssStore,
    private val engine: suspend () -> RssSyncEngine,
    private val events: FeedEventBus,
    private val preferences: StateFlow<AppPreferences>,
) : ViewModel() {
    private val mutableState = MutableStateFlow(FeedItemDetailUiState())
    val state: StateFlow<FeedItemDetailUiState> = mutableState.asStateFlow()

    init {
        viewModelScope.launch {
            events.events.collect { event ->
                if (event is FeedEvent.ItemChanged && event.item.id == "$feedId#$sortKey") {
                    mutableState.update { state ->
                        state.copy(
                            item = state.item?.copy(isRead = event.item.isRead, isFavorite = event.item.isFavorite),
                        )
                    }
                }
            }
        }
        viewModelScope.launch { load() }
    }

    private suspend fun load() {
        try {
            val item = store.item(feedId, sortKey)
            val subscription = item?.subscriptionId?.ifEmpty { null }?.let { store.subscription(it) }
            val policy =
                FeedDetailPolicy.initial(
                    subscription = subscription,
                    hasArticleUrl = FeedItemDetailUiState(item = item).articleUrl != null,
                    globalRemoteContent = preferences.value.loadRemoteContent,
                )
            mutableState.update {
                it.copy(
                    item = item,
                    subscription = subscription,
                    showingArticle = policy.showsArticle,
                    readerMode = policy.readerMode,
                    remoteContentAllowed = policy.remoteContentAllowed,
                    loaded = true,
                )
            }
            if (item != null && !item.isRead && preferences.value.effectiveRssMarkAsRead == MarkAsRead.ON_OPEN) {
                setRead(true)
            }
        } catch (exception: Exception) {
            if (exception is CancellationException) throw exception
            mutableState.update {
                it.copy(
                    error = feedUserMessage(exception, "Could not open the item"),
                    loaded = true,
                )
            }
        }
    }

    fun setRead(isRead: Boolean) {
        val item = mutableState.value.item ?: return
        val changed = item.copy(isRead = isRead, isReadExplicit = true)
        mutableState.update { it.copy(item = changed) }
        viewModelScope.launch {
            runCatching { engine().setRead(item, isRead) }
            events.post(FeedEvent.ItemChanged(changed))
        }
    }

    fun setFavorite(isFavorite: Boolean) {
        val item = mutableState.value.item ?: return
        val changed = item.copy(isFavorite = isFavorite)
        mutableState.update { it.copy(item = changed) }
        viewModelScope.launch {
            runCatching { engine().setFavorite(item, isFavorite) }
            events.post(FeedEvent.ItemChanged(changed))
        }
    }

    /** Article view on or off; writes `default_open_mode` when there is an article to open. */
    fun toggleArticle() {
        val current = mutableState.value
        val showing = !current.showingArticle
        mutableState.update { it.copy(showingArticle = showing) }
        current.subscription?.let { sub ->
            persistDefault(FeedDetailPolicy.articleUpdate(sub, showing, current.articleUrl != null))
        }
    }

    /** Reader or original styling; writes `default_styling`. */
    fun toggleReaderMode() {
        val current = mutableState.value
        val reader = !current.readerMode
        mutableState.update { it.copy(readerMode = reader) }
        current.subscription?.let { persistDefault(FeedDetailPolicy.stylingUpdate(it, reader)) }
    }

    /** Remote content on or off for this feed; writes an explicit `default_remote_content`. */
    fun toggleRemoteContent() {
        val current = mutableState.value
        val allowed = !current.remoteContentAllowed
        mutableState.update { it.copy(remoteContentAllowed = allowed) }
        current.subscription?.let { persistDefault(FeedDetailPolicy.remoteContentUpdate(it, allowed)) }
    }

    /**
     * Holds the updated row at once (so a second toggle compares against
     * the first write) and pushes it; a failure leaves the reader's state
     * alone — the next catalog refresh reconciles the row.
     */
    private fun persistDefault(update: RssSubscriptionUpdate?) {
        if (update == null) return
        val subscription = mutableState.value.subscription ?: return
        mutableState.update { it.copy(subscription = subscription.applying(update)) }
        viewModelScope.launch {
            runCatching { engine().updateSubscription(subscription, update) }
                .onSuccess { updated -> mutableState.update { it.copy(subscription = updated) } }
            events.post(FeedEvent.CatalogChanged)
        }
    }

    companion object {
        fun factory(
            container: AppContainer,
            feedId: String,
            sortKey: String,
        ): ViewModelProvider.Factory =
            viewModelFactory {
                initializer {
                    FeedItemDetailViewModel(
                        feedId = feedId,
                        sortKey = sortKey,
                        store = container.rssStore,
                        engine = { container.requireRssSync() },
                        events = container.feedEvents,
                        preferences = container.preferences.preferences,
                    )
                }
            }
    }
}
